### ISSUE TITLE
Split FEProblemBase.C and Moose.C

### ISSUE THREAD

**Issue by friedmud** [idaholab/moose#10914]
## Rationale
`FEProblemBase.C` and `Moose.C` take a prohibitively long time to compile... 50-60 seconds each.  They are holding up our compile times at this point.

## Description
I would like to split them up.

I propose taking Moose.C and splitting it up so that there is one "registration" function in each subdirectory (kernels, bcs, dirackernels, etc.) that `#includes` the stuff in that directory (this will work perfectly with Unity build too).

For FEProblemBase.C I plan on just splitting it into multiple .C files based on capability - again trying to split the includes up in a logical way.  This might be a first good step toward eventually modularizing FEProblemBase (this will at least identify logical breaking points.)

Note: in general I don't like breaking .C files into multiple files like this.  It makes it difficult to remember what piece is in what file... but in this case we don't have much of a choice.

## Impact
Faster compiles

**Comment by permcody** [comment]
To be clear, are you only proposing on breaking up Moose.C (aside from FEProblem), but not propagating that pattern into the modules and apps? I ask because it will unfortunately make MOOSE inconsistent from everything else.

**Comment by dschwen** [comment]
So... wait... with unity builds we speed up compile times by lumping stuff together (and it works impressively well). But now you want to _split_ stuff up to make it compile faster?!

**Comment by friedmud** [comment]
@dschwen Yes: split stuff _into_ the lumps.  The parts of `Moose.C` that `#include` each system will go into the directory with each system... so it will get compiled as part of the unity build of that system.

@permcody  I wasn't planning to propagate that pattern to all the other apps.  It adds unnecessary bookkeeping when the apps are pretty small.  I agree that I like uniformity... but sometimes there are reasons to break it.

If we end up really liking this for organizational purposes and we want to propagate it to the apps that's fine by me.  Let's see what it looks like first (and what we gain by it).

**Comment by friedmud** [comment]
@dschwen As for `FEProblemBase.C` the problem there is that the file is too damn big (it's really a code smell - but we can't fix it at the moment).  I want to split it up so that it can be built in parallel.

It currently takes ~50 seconds or so just to compile `FEProblemBase.C`.  I'm hoping to get that down to ~five 10-second compiles or so.
